onadata/libs/serializers/data_serializer.py

```diff
@@ -113,3 +146,87 @@ def to_representation(self, obj):
         return {
             'url': url, 'title': title, 'id_string': id_string, 'user': user
         }
+
+
+class JSONSubmissionSerializer(SubmissionSuccessMixin, serializers.Serializer):
+    def validate(self, attrs):
+        request = self.context['request']
+
+        if 'submission' not in request.data:
+            raise serializers.ValidationError({
+                'submission': _(u"No submission key provided.")
+            })
+
+        submission = request.data.get('submission')
+        if not submission:
+            raise serializers.ValidationError({
+                'submission': _(u"Received empty submission. No instance was created")
+            })
+        return super(JSONSubmissionSerializer, self).validate(attrs)
+
+    def create(self, validated_data):
+        request = self.context['request']
+        view = self.context['view']
+
+        username = view.kwargs.get('username')
+        if not username:
+            # get the username from the user if not set
+            username = (request.user and request.user.username)
+
+        submission = request.data.get('submission')
+        # convert lists in submission dict to joined strings
+        try:
+            submission_joined = dict_paths2dict(dict_lists2strings(submission))
+        except AttributeError:
+            raise serializers.ValidationError(_(u'Incorrect format, see format details here,'
+                                                u'https://api.ona.io/static/docs/submissions.html.'))
+        xml_string = dict2xform(submission_joined, request.data.get('id'))
+
+        xml_file = StringIO.StringIO(xml_string)
+
+        error, instance = safe_create_instance(
+            username, xml_file, [], None, request)
+        if error and error.status_code >= 400:
```

Comment:
why does this care about `error.status_code`, but neither the normal submission nor the RapidPro submission do?

code from 183-192 and 224-232 are the same except for the above, looks like it could be refactored into a helper

Reply:
For JSON submission we need to return a JSON response, we don't need to that for RapidPro or XML Submission. `error` is an actual `HTTPResponse` object, and in this other cases we simply append the response to `APIException` class, which we look for in the `XFormSubmissionViewSet.handle_exceptiion(exc)`. As such, the handling is different. This struck the right balance of maintaining legacy code without getting into a lot of re-writing at this time.

Reply:
Makes sense, there are cases when an error is created with a `< 400` status code?